File: packages/go/rules.go
package oleoo

import (
	"bytes"
	_ "embed"
	"encoding/json"
	"fmt"
)

//go:generate cp ../../rules.json rules.json
//go:embed rules.json
var rulesJSON []byte

// A rule is a pattern, or { pattern, notAfter } when the match must not follow notAfter: it stands in for a lookbehind.
type rule struct {
	pattern  string
	notAfter string
	str      bool
}

func (r *rule) UnmarshalJSON(b []byte) error {
	if len(b) > 0 && b[0] == '"' {
		r.str = true
		return json.Unmarshal(b, &r.pattern)
	}

	var o struct {
		Pattern  string `json:"pattern"`
		NotAfter string `json:"notAfter"`
	}
	err := json.Unmarshal(b, &o)
	r.pattern, r.notAfter = o.Pattern, o.NotAfter
	return err
}

type ruleKey struct {
	key   string
	rules []rule
}

// ruleSet keeps the keys in file order: when several keys match, the last declared one wins.
type ruleSet []ruleKey

func (s *ruleSet) UnmarshalJSON(b []byte) error {
	dec := json.NewDecoder(bytes.NewReader(b))

	if _, err := dec.Token(); err != nil {
		return err
	}

	for dec.More() {
		token, err := dec.Token()
		if err != nil {
			return err
		}

		key, ok := token.(string)
		if !ok {
			return fmt.Errorf("expected a key, got %v", token)
		}

		var rules []rule
		if err := dec.Decode(&rules); err != nil {
			return err
		}
		*s = append(*s, ruleKey{key, rules})
	}

	return nil
}

// entry is a flag of stringify.after*: a flag name, or { flag, dubRelated }.
type entry struct {
	flag       string
	dubRelated bool
	object     bool
}

func (e *entry) UnmarshalJSON(b []byte) error {
	if len(b) > 0 && b[0] == '"' {
		return json.Unmarshal(b, &e.flag)
	}

	var o struct {
		Flag       string `json:"flag"`
		DubRelated bool   `json:"dubRelated"`
	}
	err := json.Unmarshal(b, &o)
	e.flag, e.dubRelated, e.object = o.Flag, o.DubRelated, true
	return err
}

type ruleFile struct {
	Source     ruleSet  `json:"source"`
	Encoding   ruleSet  `json:"encoding"`
	Resolution ruleSet  `json:"resolution"`
	Dub        ruleSet  `json:"dub"`
	Language   ruleSet  `json:"language"`
	Flags      ruleSet  `json:"flags"`
	Erase      []string `json:"erase"`
	Extensions []string `json:"extensions"`
	Ambiguous  struct {
		Flags    []string `json:"flags"`
		Patterns []string `json:"patterns"`
	} `json:"ambiguous"`
	Title struct {
		Uppercase    []string `json:"uppercase"`
		LeadingFlags []string `json:"leadingFlags"`
		Franchises   []string `json:"franchises"`
		LeadingYears []struct {
			Year     string `json:"year"`
			Contains string `json:"contains"`
			Release  string `json:"release"`
		} `json:"leadingYears"`
	} `json:"title"`
	Stringify struct {
		DubRelated      []string `json:"dubRelated"`
		AfterTitle      []entry  `json:"afterTitle"`
		AfterYear       []entry  `json:"afterYear"`
		AfterLanguage   []entry  `json:"afterLanguage"`
		AfterResolution []entry  `json:"afterResolution"`
		AfterSource     []entry  `json:"afterSource"`
		AfterEncoding   []entry  `json:"afterEncoding"`
		AfterDub        []entry  `json:"afterDub"`
	} `json:"stringify"`
}

// A package-level initializer, not init(): the regexes of oleoo.go read the rules when they are declared.
var rules = func() (r ruleFile) {
	if err := json.Unmarshal(rulesJSON, &r); err != nil {
		panic("oleoo: rules.json: " + err.Error())
	}
	return r
}()

File: packages/rust/src/rules.rs
use std::sync::LazyLock;

use serde::{Deserialize, Deserializer, de::Error as _};
use serde_json::{Map, Value};

pub static RULES: LazyLock<Rules> = LazyLock::new(|| {
    serde_json::from_str(include_str!(concat!(env!("OUT_DIR"), "/rules.json")))
        .expect("rules.json does not follow SPEC.md")
});

// A rule is a pattern, or { pattern, notAfter } when the match must not follow notAfter: it stands in for a lookbehind.
#[derive(Deserialize)]
#[serde(untagged)]
pub enum Rule {
    Pattern(String),
    Guarded {
        pattern: String,
        #[serde(rename = "notAfter")]
        not_after: String,
    },
}

impl Rule {
    pub fn pattern(&self) -> &str {
        match self {
            Rule::Pattern(pattern) | Rule::Guarded { pattern, .. } => pattern,
        }
    }

    pub fn not_after(&self) -> Option<&str> {
        match self {
            Rule::Pattern(_) => None,
            Rule::Guarded { not_after, .. } => Some(not_after),
        }
    }

    pub fn anchored(&self) -> bool {
        matches!(self, Rule::Pattern(pattern) if pattern.starts_with('^'))
    }
}

// The keys of source, encoding, resolution, dub, language and flags, in file order: the last matching key wins.
pub type Keys = Vec<(String, Vec<Rule>)>;

#[derive(Deserialize)]
pub struct Rules {
    #[serde(deserialize_with = "ordered")]
    pub source: Keys,
    #[serde(deserialize_with = "ordered")]
    pub encoding: Keys,
    #[serde(deserialize_with = "ordered")]
    pub resolution: Keys,
    #[serde(deserialize_with = "ordered")]
    pub dub: Keys,
    #[serde(deserialize_with = "ordered")]
    pub language: Keys,
    #[serde(deserialize_with = "ordered")]
    pub flags: Keys,
    pub erase: Vec<String>,
    pub extensions: Vec<String>,
    pub ambiguous: Ambiguous,
    pub title: Title,
    pub stringify: Order,
}

#[derive(Deserialize)]
pub struct Ambiguous {
    pub flags: Vec<String>,
    pub patterns: Vec<String>,
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Title {
    pub uppercase: Vec<String>,
    pub leading_flags: Vec<String>,
    pub franchises: Vec<String>,
    pub leading_years: Vec<LeadingYear>,
}

#[derive(Deserialize)]
pub struct LeadingYear {
    pub year: String,
    pub contains: String,
    pub release: String,
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Order {
    pub dub_related: Vec<String>,
    pub after_title: Vec<Entry>,
    pub after_year: Vec<Entry>,
    pub after_language: Vec<Entry>,
    pub after_resolution: Vec<Entry>,
    pub after_source: Vec<Entry>,
    pub after_encoding: Vec<Entry>,
    pub after_dub: Vec<Entry>,
}

#[derive(Deserialize)]
#[serde(untagged)]
pub enum Entry {
    Flag(String),
    Placed {
        flag: String,
        #[serde(rename = "dubRelated")]
        dub_related: bool,
    },
}

fn ordered<'de, D: Deserializer<'de>>(deserializer: D) -> Result<Keys, D::Error> {
    Map::<String, Value>::deserialize(deserializer)?
        .into_iter()
        .map(|(key, rules)| {
            Ok((
                key,
                serde_json::from_value(rules).map_err(D::Error::custom)?,
            ))
        })
        .collect()
}
